contracts/FlowIDTableStaking.cdc

```diff
@@ -833,41 +875,55 @@ pub contract FlowIDTableStaking {
-                    // since they are forced to unstake
-                    nodeRecord.tokensRequestedToUnstake = nodeRecord.tokensStaked.balance
-
-                    // Iterate through all delegators and unstake their tokens
-                    // since their node has unstaked
-                    for delegator in nodeRecord.delegators.keys {
-                        let delRecord = nodeRecord.borrowDelegatorRecord(delegator)
-
-                        if delRecord.tokensCommitted.balance > 0.0 {
-                            emit DelegatorTokensUnstaked(nodeID: nodeRecord.id, delegatorID: delegator, amount: delRecord.tokensCommitted.balance)
-
-                            // move their committed tokens back to their unstaked tokens
-                            delRecord.tokensUnstaked.deposit(from: <-delRecord.tokensCommitted.withdraw(amount: delRecord.tokensCommitted.balance))
-                        }
-
-                        // Request to unstake all tokens
-                        delRecord.tokensRequestedToUnstake = delRecord.tokensStaked.balance
+                   (nodeRecord.role != UInt8(5) && approvedNodeIDs[nodeID] == nil) {
+                    self.removeAndRefundNodeRecord(nodeRecord: nodeRecord)
+                } else {
+                    //capture whether this is a current node or a pending node
+                    if nodeRecord.initialWeight > 0 {
+                        currentNodeCount[nodeRecord.role] = currentNodeCount[nodeRecord.role]! + 1
+                    } else if candidateNodes!.contains(nodeID) {
+                        pendingNodes[nodeRecord.role]!.append(nodeRecord)
                     }
 
-                    // Clear initial weight because the node is not staked any more
-                    nodeRecord.initialWeight = 0
-                } else {
                     // Set weight to 100
                     // Calculations for node weight will come with a future version of epochs
                     nodeRecord.initialWeight = 100
                 }
             }
+
+            let slotLimits = FlowIDTableStaking.account.borrow<&{UInt8: UInt16}>(from: /storage/flowStakingSlotLimits)
+                ?? panic("Could not load the slot limits object from storage!")
+
+            for nodeRole in currentNodeCount.keys {
+                if currentNodeCount[nodeRole]! >= slotLimits[nodeRole]! {
+                    //if all slots are full, remove and refund all pending nodes
+                    for nodeRecord in pendingNodes[nodeRole]! {
+                        self.removeAndRefundNodeRecord(nodeRecord: nodeRecord)
+                    }
+                } else if currentNodeCount[nodeRole]! + UInt16(pendingNodes[nodeRole]!.length) > slotLimits[nodeRole]! {
+                    //randomly remove pending nodes until the limit is met
+                    let deletionCount: UInt16 = currentNodeCount[nodeRole]! + UInt16(pendingNodes[nodeRole]!.length) - slotLimits[nodeRole]!
```

Comment:
I think that's an attack here because of this method of deleting nodes given that currently there are likely to be few free slots.


Imagine:
nodeCount[nodeRole] == 5
slotLimits[nodeRole] == 6
pendingNodes[nodeRole] == 10000

This could happen by someone simply generating 10000 keys and doing the node registration transaction for them.

Now, when this function runs we need to reject 9999 nodes in order to fill 1 slot.

Instead, we should be choosing nodes to include. I understand that the above method gets around the duplicate issue. The options I see:

1) Track chosen nodes in a dictionary by id (such as my original example)
2) Scan the selected array for duplicates upon adding each.
3) Use "in place swap" method for choosing a random subset

Given the reality of the number of open slots I think (1) is fine.

Reply:
@Wolog2021 Pointed out that no matter what we need to go through all 10000 nodes in the above example, because the code needs to remove and refund those tokens. So you still always need to touch all nodes.

DUH